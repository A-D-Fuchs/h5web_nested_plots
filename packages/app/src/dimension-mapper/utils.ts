import type { Axis } from '@h5web/shared/vis-models';

export function isAxis(elem: number | Axis): elem is Axis {
  return typeof elem !== 'number';
}
