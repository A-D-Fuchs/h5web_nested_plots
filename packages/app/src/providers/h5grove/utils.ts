import type {
  Attribute,
  ChildEntity,
  DType,
  Group,
  ProvidedEntity,
} from '@h5web/shared/hdf5-models';
import { Endianness, EntityKind } from '@h5web/shared/hdf5-models';
import {
  boolType,
  buildEntityPath,
  compoundType,
  cplxType,
  floatType,
  intType,
  strType,
  uintType,
  unknownType,
} from '@h5web/shared/hdf5-utils';

import type { H5GroveAttribute, H5GroveDtype, H5GroveEntity } from './models';

export function parseEntity(
  path: string,
  h5gEntity: H5GroveEntity,
  isChild: true,
): ChildEntity;

export function parseEntity(
  path: string,
  h5gEntity: H5GroveEntity,
  isChild?: false,
): ProvidedEntity;

export function parseEntity(
  path: string,
  h5gEntity: H5GroveEntity,
  isChild = false,
): ProvidedEntity | ChildEntity {
  const { name } = h5gEntity;
  const baseEntity = { name, path };

  if (h5gEntity.type === EntityKind.Group) {
    const { children = [], attributes: attrsMetadata } = h5gEntity;
    const attributes = parseAttributes(attrsMetadata);
    const baseGroup: Group = {
      ...baseEntity,
      kind: EntityKind.Group,
      attributes,
    };

    if (isChild) {
      return baseGroup;
    }

    return {
      ...baseGroup,
      children: children.map((child) => {
        const childPath = buildEntityPath(path, child.name);
        return parseEntity(childPath, child, true);
      }),
    };
  }

  if (h5gEntity.type === EntityKind.Dataset) {
    const {
      attributes: attrsMetadata,
      dtype,
      shape,
      chunks,
      filters,
    } = h5gEntity;
    return {
      ...baseEntity,
      kind: EntityKind.Dataset,
      shape,
      type: parseDType(dtype),
      rawType: dtype,
      ...(chunks && { chunks }),
      ...(filters && { filters }),
      attributes: parseAttributes(attrsMetadata),
    };
  }

  if (h5gEntity.type === 'soft_link') {
    const { target_path } = h5gEntity;
    return {
      ...baseEntity,
      kind: EntityKind.Unresolved,
      attributes: [],
      link: { class: 'Soft', path: target_path },
    };
  }

  if (h5gEntity.type === 'external_link') {
    const { target_file, target_path } = h5gEntity;
    return {
      ...baseEntity,
      kind: EntityKind.Unresolved,
      attributes: [],
      link: {
        class: 'External',
        file: target_file,
        path: target_path,
      },
    };
  }

  // Treat other entities as unresolved
  return {
    ...baseEntity,
    kind: EntityKind.Unresolved,
    attributes: [],
  };
}

function parseAttributes(attrsMetadata: H5GroveAttribute[]): Attribute[] {
  return attrsMetadata.map<Attribute>(({ name, dtype, shape }) => ({
    name,
    shape,
    type: parseDType(dtype),
  }));
}

// https://numpy.org/doc/stable/reference/generated/numpy.dtype.byteorder.html#numpy.dtype.byteorder
const ENDIANNESS_MAPPING: Record<string, Endianness> = {
  '<': Endianness.LE,
  '>': Endianness.BE,
};

export function parseDType(dtype: H5GroveDtype): DType {
  if (typeof dtype === 'string') {
    return parseDTypeFromString(dtype);
  }

  return compoundType(
    Object.fromEntries(
      Object.entries(dtype).map(([k, v]) => [k, parseDType(v)]),
    ),
  );
}

function parseDTypeFromString(dtype: string): DType {
  const regexp = /([<>=|])?([A-Za-z])(\d*)/u;
  const matches = regexp.exec(dtype);

  if (matches === null) {
    throw new Error(`Invalid dtype string: ${dtype}`);
  }

  const [, endianMatch, dataType, lengthMatch] = matches;

  const length = lengthMatch ? Number.parseInt(lengthMatch, 10) : 0;
  const endianness = ENDIANNESS_MAPPING[endianMatch] || undefined;

  switch (dataType) {
    case 'b':
      // Booleans are stored as bytes but numpy represents them distinctly from "normal" bytes:
      // `|b1` for booleans vs. `|i1` for normal bytes
      // https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.bool
      return boolType();

    case 'f':
      return floatType(length * 8, endianness);

    case 'i':
      return intType(length * 8, endianness);

    case 'u':
      return uintType(length * 8, endianness);

    case 'c':
      return cplxType(
        floatType(
          (length / 2) * 8, // bytes are equally distributed between real and imag
          endianness,
        ),
      );

    case 'S':
      return strType('ASCII', length);

    case 'O':
      return strType('UTF-8');

    default:
      return unknownType();
  }
}

export function hasErrorMessage(error: unknown): error is { message: string } {
  return !!error && typeof error === 'object' && 'message' in error;
}
