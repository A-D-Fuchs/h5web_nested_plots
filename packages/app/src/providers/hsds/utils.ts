import { isGroup } from '@h5web/shared/guards';
import type {
  ArrayShape,
  Attribute,
  ComplexType,
  CompoundType,
  Dataset,
  DType,
  Entity,
  Group,
  NumericType,
  ScalarShape,
  Shape,
} from '@h5web/shared/hdf5-models';
import { DTypeClass, Endianness } from '@h5web/shared/hdf5-models';
import {
  boolType,
  compoundType,
  cplxType,
  floatType,
  intType,
  strType,
  unknownType,
} from '@h5web/shared/hdf5-utils';

import type {
  HsdsAttribute,
  HsdsCompoundType,
  HsdsEntity,
  HsdsNumericType,
  HsdsShape,
  HsdsType,
} from './models';

export function isHsdsGroup(entity: HsdsEntity): entity is HsdsEntity<Group> {
  return isGroup(entity);
}

export function assertHsdsEntity<T extends Entity>(
  entity: T,
): asserts entity is HsdsEntity<T> {
  if (!('id' in entity)) {
    throw new Error('Expected entity to be HSDS entity');
  }
}

export function assertHsdsDataset(
  dataset: Dataset<ScalarShape | ArrayShape>,
): asserts dataset is HsdsEntity<Dataset<ScalarShape | ArrayShape>> {
  if (!('id' in dataset)) {
    throw new Error('Expected entity to be HSDS dataset');
  }
}

function assertHsdsNumericType(
  type: HsdsType,
): asserts type is HsdsNumericType {
  if (type.class !== 'H5T_INTEGER' && type.class !== 'H5T_FLOAT') {
    throw new Error('Expected HSDS numeric type');
  }
}

export function convertHsdsShape(shape: HsdsShape): Shape {
  const { class: shapeClass, dims } = shape;
  return dims || (shapeClass === 'H5S_SCALAR' ? [] : null);
}

function convertHsdsNumericType(hsdsType: HsdsNumericType): NumericType {
  const { class: hsdsClass, base } = hsdsType;

  const regex = /H5T_(?:STD|IEEE)_([A-Z])(\d+)(BE|LE)/u;
  const matches = regex.exec(base);

  if (!matches) {
    throw new Error(`Unrecognized base ${base}`);
  }

  const [, sign, sizeStr, endiannessAbbr] = matches;
  const unsigned = sign === 'U';
  const size = Number.parseInt(sizeStr, 10);
  const endianness = Endianness[endiannessAbbr as 'BE' | 'LE'];

  if (hsdsClass === 'H5T_FLOAT') {
    return floatType(size, endianness);
  }

  return intType(size, unsigned, endianness);
}

function convertHsdsCompoundType(
  hsdsType: HsdsCompoundType,
): CompoundType | ComplexType {
  const { fields } = hsdsType;

  if (fields.length === 2 && fields[0].name === 'r' && fields[1].name === 'i') {
    const [{ type: realType }, { type: imagType }] = fields;
    assertHsdsNumericType(realType);
    assertHsdsNumericType(imagType);

    return cplxType(
      convertHsdsNumericType(realType),
      convertHsdsNumericType(imagType),
    );
  }

  return compoundType(
    Object.fromEntries(
      hsdsType.fields.map((v) => [v.name, convertHsdsType(v.type)]),
    ),
  );
}

export function convertHsdsType(hsdsType: HsdsType): DType {
  switch (hsdsType.class) {
    case 'H5T_INTEGER':
    case 'H5T_FLOAT':
      return convertHsdsNumericType(hsdsType);

    case 'H5T_COMPOUND':
      return convertHsdsCompoundType(hsdsType);

    case 'H5T_STRING':
      return strType(
        hsdsType.charSet.endsWith('ASCII') ? 'ASCII' : 'UTF-8',
        hsdsType.length === 'H5T_VARIABLE' ? undefined : hsdsType.length,
      );

    case 'H5T_ARRAY':
    case 'H5T_VLEN':
      return {
        class:
          hsdsType.class === 'H5T_ARRAY' ? DTypeClass.Array : DTypeClass.VLen,
        base: convertHsdsType(hsdsType.base),
        ...(hsdsType.dims && { dims: hsdsType.dims }),
      };

    case 'H5T_ENUM':
      // Booleans are stored as Enum by h5py
      // https://docs.h5py.org/en/stable/faq.html#what-datatypes-are-supported
      if (hsdsType.mapping.FALSE === 0) {
        return boolType();
      }

      return {
        class: DTypeClass.Enum,
        base: convertHsdsType(hsdsType.base),
        mapping: hsdsType.mapping,
      };

    default:
      return unknownType();
  }
}

export function convertHsdsAttributes(attrs: HsdsAttribute[]): Attribute[] {
  return attrs.map((attr) => ({
    name: attr.name,
    shape: convertHsdsShape(attr.shape),
    type: convertHsdsType(attr.type),
  }));
}
