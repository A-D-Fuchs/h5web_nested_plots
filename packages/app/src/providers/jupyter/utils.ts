import { EntityKind } from '@h5web/shared';

import type {
  JupyterDatasetResponse,
  JupyterGroupResponse,
  JupyterEntityResponse,
} from './models';

export function isGroupResponse(
  response: JupyterEntityResponse
): response is JupyterGroupResponse {
  return response.type === EntityKind.Group;
}

export function isDatasetResponse(
  response: JupyterEntityResponse
): response is JupyterDatasetResponse {
  return response.type === EntityKind.Dataset;
}
