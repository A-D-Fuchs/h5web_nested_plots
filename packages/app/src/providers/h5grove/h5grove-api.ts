import { hasNumericType, hasScalarShape } from '@h5web/shared/guards';
import type {
  ArrayShape,
  AttributeValues,
  Dataset,
  Entity,
  ProvidedEntity,
  Value,
} from '@h5web/shared/hdf5-models';
import type { AxiosRequestConfig } from 'axios';

import { DataProviderApi } from '../api';
import type { ExportFormat, ExportURL, ValuesStoreParams } from '../models';
import { handleAxiosError } from '../utils';
import type {
  H5GroveAttrValuesResponse,
  H5GroveDataResponse,
  H5GroveEntityResponse,
  H5GrovePathsResponse,
} from './models';
import {
  h5groveTypedArrayFromDType,
  hasErrorMessage,
  parseEntity,
} from './utils';

export class H5GroveApi extends DataProviderApi {
  /* API compatible with h5grove@2.0.0b2 */
  public constructor(
    url: string,
    filepath: string,
    axiosConfig?: AxiosRequestConfig,
    private readonly _getExportURL?: DataProviderApi['getExportURL'],
  ) {
    super(filepath, { baseURL: url, ...axiosConfig });
  }

  public async getEntity(path: string): Promise<ProvidedEntity> {
    const response = await this.fetchEntity(path);
    return parseEntity(path, response);
  }

  public async getValue(
    params: ValuesStoreParams,
  ): Promise<H5GroveDataResponse> {
    const { dataset } = params;

    const DTypedArray = h5groveTypedArrayFromDType(dataset.type);
    if (DTypedArray) {
      const buffer = await this.fetchBinaryData(params);
      const array = new DTypedArray(buffer);
      return hasScalarShape(dataset) ? array[0] : array;
    }

    return this.fetchData(params);
  }

  public async getAttrValues(entity: Entity): Promise<AttributeValues> {
    const { path, attributes } = entity;
    return attributes.length > 0 ? this.fetchAttrValues(path) : {};
  }

  public override getExportURL<D extends Dataset<ArrayShape>>(
    format: ExportFormat,
    dataset: D,
    selection: string | undefined,
    value: Value<D>,
  ): ExportURL {
    const url = this._getExportURL?.(format, dataset, selection, value);
    if (url) {
      return url;
    }

    if (format !== 'json' && !hasNumericType(dataset)) {
      return undefined;
    }

    const { baseURL, params } = this.client.defaults;

    const searchParams = new URLSearchParams(params as Record<string, string>);
    searchParams.set('path', dataset.path);
    searchParams.set('format', format);

    if (selection) {
      searchParams.set('selection', selection);
    }

    return new URL(`${baseURL as string}/data/?${searchParams.toString()}`);
  }

  public override async getSearchablePaths(path: string): Promise<string[]> {
    const { data } = await this.client.get<H5GrovePathsResponse>(`/paths/`, {
      params: { path },
    });

    return data;
  }

  private async fetchEntity(path: string): Promise<H5GroveEntityResponse> {
    const { data } = await handleAxiosError(
      () =>
        this.client.get<H5GroveEntityResponse>(`/meta/`, { params: { path } }),
      (_, errorData) => {
        if (!hasErrorMessage(errorData)) {
          return undefined;
        }
        const { message } = errorData;

        if (message.includes('File not found')) {
          return `File not found: '${this.filepath}'`;
        }
        if (message.includes('Permission denied')) {
          return `Cannot read file '${this.filepath}': Permission denied`;
        }
        if (message.includes('not a valid path')) {
          return `No entity found at ${path}`;
        }
        if (message.includes('Cannot resolve')) {
          return `Could not resolve soft link at ${path}`;
        }

        return undefined;
      },
    );
    return data;
  }

  private async fetchAttrValues(
    path: string,
  ): Promise<H5GroveAttrValuesResponse> {
    const { data } = await this.client.get<H5GroveAttrValuesResponse>(
      `/attr/`,
      { params: { path } },
    );
    return data;
  }

  private async fetchData(
    params: ValuesStoreParams,
  ): Promise<H5GroveDataResponse> {
    const { data } = await this.cancellableFetchValue(`/data/`, params, {
      path: params.dataset.path,
      selection: params.selection,
      flatten: true,
    });

    return data;
  }

  private async fetchBinaryData(
    params: ValuesStoreParams,
  ): Promise<ArrayBuffer> {
    const { data } = await this.cancellableFetchValue(
      '/data/',
      params,
      {
        path: params.dataset.path,
        selection: params.selection,
        format: 'bin',
        dtype: 'safe',
      },
      'arraybuffer',
    );

    return data;
  }
}
