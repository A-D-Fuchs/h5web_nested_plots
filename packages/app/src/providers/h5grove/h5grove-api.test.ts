import {
  assertDataset,
  assertGroup,
  assertGroupWithChildren,
  hasNonNullShape,
} from '@h5web/shared/guards';
import { expect, test } from 'vitest';

import { getValueOrError } from '../utils';
import { H5GroveApi } from './h5grove-api';

const H5GROVE_URL = 'http://localhost:8888'; // when running `pnpm support:h5grove`
const TEST_FILE = 'sample.h5';

test('test file matches snapshot', async () => {
  const api = new H5GroveApi(H5GROVE_URL, TEST_FILE, {
    params: { file: TEST_FILE },
  });

  const root = await api.getEntity('/');
  assertGroup(root);
  assertGroupWithChildren(root);

  const children = await Promise.all(
    root.children.map(async (child) => {
      assertDataset(child);
      const { name, shape, type, rawType } = child;

      const value = hasNonNullShape(child)
        ? await getValueOrError(api, child)
        : null;

      return { name, shape, type, rawType, value };
    }),
  );

  expect(children).toMatchSnapshot();
});
