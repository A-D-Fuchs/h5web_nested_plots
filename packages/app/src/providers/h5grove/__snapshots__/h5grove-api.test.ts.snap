// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test file matches snapshot 1`] = `
[
  {
    "name": "int8_scalar",
    "rawType": "|i1",
    "shape": [],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 8,
    },
    "value": -128,
  },
  {
    "name": "int8_2D",
    "rawType": "|i1",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 8,
    },
    "value": Int8Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "int16_scalar",
    "rawType": "<i2",
    "shape": [],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 16,
    },
    "value": -32768,
  },
  {
    "name": "int16_2D",
    "rawType": "<i2",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 16,
    },
    "value": Int16Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "int32_scalar",
    "rawType": "<i4",
    "shape": [],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": -2147483648,
  },
  {
    "name": "int32_BE_scalar",
    "rawType": ">i4",
    "shape": [],
    "type": {
      "class": "Integer",
      "endianness": "big-endian",
      "size": 32,
    },
    "value": 0,
  },
  {
    "name": "int32_2D",
    "rawType": "<i4",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": Int32Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "int64_scalar",
    "rawType": "<i8",
    "shape": [],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": -9223372036854776000,
  },
  {
    "name": "int64_2D",
    "rawType": "<i8",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "uint8_scalar",
    "rawType": "|u1",
    "shape": [],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 8,
    },
    "value": 255,
  },
  {
    "name": "uint8_2D",
    "rawType": "|u1",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 8,
    },
    "value": Uint8Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "uint16_scalar",
    "rawType": "<u2",
    "shape": [],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 16,
    },
    "value": 65535,
  },
  {
    "name": "uint16_2D",
    "rawType": "<u2",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 16,
    },
    "value": Uint16Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "uint32_scalar",
    "rawType": "<u4",
    "shape": [],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": 4294967295,
  },
  {
    "name": "uint32_2D",
    "rawType": "<u4",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": Uint32Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "uint64_scalar",
    "rawType": "<u8",
    "shape": [],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": 18446744073709552000,
  },
  {
    "name": "uint64_3D",
    "rawType": "<u8",
    "shape": [
      2,
      2,
      2,
    ],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      18446744073709552000,
    ],
  },
  {
    "name": "float16_scalar",
    "rawType": "<f2",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 16,
    },
    "value": 0.00006103515625,
  },
  {
    "name": "float16_2D",
    "rawType": "<f2",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 16,
    },
    "value": [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "float32_empty",
    "rawType": "<f4",
    "shape": null,
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": null,
  },
  {
    "name": "float32_scalar",
    "rawType": "<f4",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": 1.1754943508222875e-38,
  },
  {
    "name": "float32_BE_scalar",
    "rawType": ">f4",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "big-endian",
      "size": 32,
    },
    "value": 0,
  },
  {
    "name": "float32_2D",
    "rawType": "<f4",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": Float32Array [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "float64_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": 2.2250738585072014e-308,
  },
  {
    "name": "float64_nan_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": NaN,
  },
  {
    "name": "float64_inf_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": Infinity,
  },
  {
    "name": "float64_ninf_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": -Infinity,
  },
  {
    "name": "float64_zero_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": 0,
  },
  {
    "name": "float64_nzero_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": -0,
  },
  {
    "name": "float64_pi_scalar",
    "rawType": "<f8",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": 3.141592653589793,
  },
  {
    "name": "float64_2D",
    "rawType": "<f8",
    "shape": [
      3,
      5,
    ],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 64,
    },
    "value": Float64Array [
      0,
      1,
      Infinity,
      3,
      4,
      3,
      4,
      NaN,
      6,
      7,
      6,
      7,
      -Infinity,
      9,
      10,
    ],
  },
  {
    "name": "float128_scalar",
    "rawType": "<f16",
    "shape": [],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 128,
    },
    "value": 0,
  },
  {
    "name": "float128_2D",
    "rawType": "<f16",
    "shape": [
      2,
      3,
    ],
    "type": {
      "class": "Float",
      "endianness": "little-endian",
      "size": 128,
    },
    "value": [
      0,
      1,
      2,
      3,
      4,
      5,
    ],
  },
  {
    "name": "ascii_vlen_empty",
    "rawType": "|O",
    "shape": null,
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": null,
  },
  {
    "name": "ascii_vlen_scalar",
    "rawType": "|O",
    "shape": [],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": "Some text",
  },
  {
    "name": "ascii_fixed_scalar",
    "rawType": "|S9",
    "shape": [],
    "type": {
      "charSet": "ASCII",
      "class": "String",
      "length": 9,
    },
    "value": "Some text",
  },
  {
    "name": "utf8_vlen_scalar",
    "rawType": "|O",
    "shape": [],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": "Some text",
  },
  {
    "name": "utf8_vlen_1D",
    "rawType": "|O",
    "shape": [
      3,
    ],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": [
      "foo",
      "bar",
      "baz",
    ],
  },
  {
    "name": "utf8_fixed_scalar",
    "rawType": "|S9",
    "shape": [],
    "type": {
      "charSet": "ASCII",
      "class": "String",
      "length": 9,
    },
    "value": "Some text",
  },
  {
    "name": "byte_string_scalar",
    "rawType": "|V3",
    "shape": [],
    "type": {
      "class": "Unknown",
    },
    "value": " "",
  },
  {
    "name": "datetime64_scalar",
    "rawType": "|V8",
    "shape": [],
    "type": {
      "class": "Unknown",
    },
    "value": [AxiosError: Request failed with status code 500],
  },
  {
    "name": "datetime64_not-a-time_scalar",
    "rawType": "|V8",
    "shape": [],
    "type": {
      "class": "Unknown",
    },
    "value": [AxiosError: Request failed with status code 500],
  },
  {
    "name": "complex64_scalar",
    "rawType": "<c8",
    "shape": [],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 32,
      },
      "realType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 32,
      },
    },
    "value": [
      1.1754943508222875e-38,
      3.4028234663852886e+38,
    ],
  },
  {
    "name": "complex64_2D",
    "rawType": "<c8",
    "shape": [
      2,
      2,
    ],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 32,
      },
      "realType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 32,
      },
    },
    "value": [
      [
        1,
        2,
      ],
      [
        3,
        4,
      ],
      [
        5,
        6,
      ],
      [
        7,
        8,
      ],
    ],
  },
  {
    "name": "complex128_scalar",
    "rawType": "<c16",
    "shape": [],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 64,
      },
      "realType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 64,
      },
    },
    "value": [
      2.2250738585072014e-308,
      1.7976931348623157e+308,
    ],
  },
  {
    "name": "complex128_BE_scalar",
    "rawType": ">c16",
    "shape": [],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "big-endian",
        "size": 64,
      },
      "realType": {
        "class": "Float",
        "endianness": "big-endian",
        "size": 64,
      },
    },
    "value": [
      1,
      2,
    ],
  },
  {
    "name": "complex128_2D",
    "rawType": "<c16",
    "shape": [
      2,
      2,
    ],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 64,
      },
      "realType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 64,
      },
    },
    "value": [
      [
        1,
        2,
      ],
      [
        3,
        4,
      ],
      [
        5,
        6,
      ],
      [
        7,
        8,
      ],
    ],
  },
  {
    "name": "complex256_scalar",
    "rawType": "<c32",
    "shape": [],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 128,
      },
      "realType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 128,
      },
    },
    "value": [AxiosError: Request failed with status code 500],
  },
  {
    "name": "complex256_2D",
    "rawType": "<c32",
    "shape": [
      2,
      2,
    ],
    "type": {
      "class": "Complex",
      "imagType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 128,
      },
      "realType": {
        "class": "Float",
        "endianness": "little-endian",
        "size": 128,
      },
    },
    "value": [AxiosError: Request failed with status code 500],
  },
  {
    "name": "compound_scalar",
    "rawType": {
      "bigint": "<i8",
      "double": "<f8",
      "utf-8": "|O",
    },
    "shape": [],
    "type": {
      "class": "Compound",
      "fields": {
        "bigint": {
          "class": "Integer",
          "endianness": "little-endian",
          "size": 64,
        },
        "double": {
          "class": "Float",
          "endianness": "little-endian",
          "size": 64,
        },
        "utf-8": {
          "charSet": "UTF-8",
          "class": "String",
        },
      },
    },
    "value": [
      1,
      2,
      "foo",
    ],
  },
  {
    "name": "compound_1D",
    "rawType": {
      "bigint": "<i8",
      "double": "<f8",
      "utf-8": "|O",
    },
    "shape": [
      3,
    ],
    "type": {
      "class": "Compound",
      "fields": {
        "bigint": {
          "class": "Integer",
          "endianness": "little-endian",
          "size": 64,
        },
        "double": {
          "class": "Float",
          "endianness": "little-endian",
          "size": 64,
        },
        "utf-8": {
          "charSet": "UTF-8",
          "class": "String",
        },
      },
    },
    "value": [
      [
        1,
        null,
        "foo",
      ],
      [
        2,
        null,
        "bar",
      ],
      [
        3,
        -0,
        "baz",
      ],
    ],
  },
  {
    "name": "compound_nested_scalar",
    "rawType": {
      "nested": {
        "bool": "|b1",
        "cplx": "<c8",
      },
    },
    "shape": [],
    "type": {
      "class": "Compound",
      "fields": {
        "nested": {
          "class": "Compound",
          "fields": {
            "bool": {
              "class": "Boolean",
            },
            "cplx": {
              "class": "Complex",
              "imagType": {
                "class": "Float",
                "endianness": "little-endian",
                "size": 32,
              },
              "realType": {
                "class": "Float",
                "endianness": "little-endian",
                "size": 32,
              },
            },
          },
        },
      },
    },
    "value": [
      [
        true,
        [
          1,
          2,
        ],
      ],
    ],
  },
  {
    "name": "compound_array_vlen_1D",
    "rawType": {
      "arr": "|V8",
      "vlen": "|O",
    },
    "shape": [
      3,
    ],
    "type": {
      "class": "Compound",
      "fields": {
        "arr": {
          "class": "Unknown",
        },
        "vlen": {
          "charSet": "UTF-8",
          "class": "String",
        },
      },
    },
    "value": [
      [
        [
          0,
          1,
        ],
        [
          0,
        ],
      ],
      [
        [
          2,
          3,
        ],
        [
          0,
          1,
        ],
      ],
      [
        [
          4,
          5,
        ],
        [
          0,
          1,
          2,
        ],
      ],
    ],
  },
  {
    "name": "reference_scalar",
    "rawType": "|O",
    "shape": [],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": "<HDF5 object reference>",
  },
  {
    "name": "reference_region_scalar",
    "rawType": "|O",
    "shape": [],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": "<HDF5 region reference>",
  },
  {
    "name": "bool_empty",
    "rawType": "|b1",
    "shape": null,
    "type": {
      "class": "Boolean",
    },
    "value": null,
  },
  {
    "name": "bool_false_scalar",
    "rawType": "|b1",
    "shape": [],
    "type": {
      "class": "Boolean",
    },
    "value": false,
  },
  {
    "name": "bool_true_scalar",
    "rawType": "|b1",
    "shape": [],
    "type": {
      "class": "Boolean",
    },
    "value": true,
  },
  {
    "name": "bool_2D",
    "rawType": "|b1",
    "shape": [
      2,
      4,
    ],
    "type": {
      "class": "Boolean",
    },
    "value": [
      true,
      false,
      true,
      true,
      false,
      false,
      true,
      false,
    ],
  },
  {
    "name": "enum_uint8_scalar",
    "rawType": "|u1",
    "shape": [],
    "type": {
      "class": "Integer (unsigned)",
      "endianness": "little-endian",
      "size": 8,
    },
    "value": 1,
  },
  {
    "name": "enum_int32_scalar",
    "rawType": "<i4",
    "shape": [],
    "type": {
      "class": "Integer",
      "endianness": "little-endian",
      "size": 32,
    },
    "value": 256,
  },
  {
    "name": "vlen_int8_scalar",
    "rawType": "|O",
    "shape": [],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": [
      0,
      1,
    ],
  },
  {
    "name": "vlen_int64_1D",
    "rawType": "|O",
    "shape": [
      3,
    ],
    "type": {
      "charSet": "UTF-8",
      "class": "String",
    },
    "value": [
      [
        0,
      ],
      [
        0,
        1,
      ],
      [
        0,
        1,
        2,
      ],
    ],
  },
]
`;
